File: leetcode/find-median-from-data-stream.cpp
/*
 * @Author: Tairan Gao
 * @Date:   2023-08-18 10:51:27
 * @Last Modified by:   Tairan Gao
 * @Last Modified time: 2023-08-18 11:21:38
 */

#include <limits.h>
#include <tuple>
#include <vector>
#include <iostream>
#include <deque>
#include <set>

using namespace std;

class MedianFinder
{
public:
    MedianFinder() : left(list.cend()), right(list.cbegin()){};

    void addNum(int num)
    {
        if (list.empty())
        {
            left = right = list.insert(num);
            return;
        }
        list.insert(num);
        size_t n = list.size();
        if (n & 1) // odd
        {
            if (num >= *right)
            {
                left = right;
            }
            else
            {
                left = --right;
            }
        }
        else
        {
            if (num >= *right)
            {
                right++;
            }
            else
            {
                left--;
            }
        }
    }

    double findMedian()
    {
        return static_cast<double>(*left + *right) / 2.0;
    }

private:
    multiset<int> list;
    multiset<int>::iterator left;
    multiset<int>::iterator right;
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */

int main()
{
    MedianFinder *obj = new MedianFinder();
    obj->addNum(1);
    obj->addNum(2);
    double param_2 = obj->findMedian();
    std::cout << param_2 << std::endl;
    obj->addNum(3);
    param_2 = obj->findMedian();
    std::cout << param_2 << std::endl;
    return 0;
}

File: leetcode/find-median-from-data-stream.py
# -*- coding: utf-8 -*-
# @Author: Tairan Gao
# @Date:   2023-08-20 22:16:09
# @Last Modified by:   Tairan Gao
# @Last Modified time: 2023-08-21 13:52:37

import heapq


class MedianFinder:
    def __init__(self):
        self.upper = []
        self.lower = []

    def addNum(self, num: int) -> None:
        if len(self.upper) == 0:
            heapq.heappush(self.upper, num)
            return

        if len(self.lower) == 0:
            if num <= self.upper[0]:
                heapq.heappush(self.lower, -num)
            else:
                exchange = heapq.heappop(self.upper)
                heapq.heappush(self.lower, -exchange)
                heapq.heappush(self.upper, num)
            return

        if len(self.upper) <= len(self.lower):
            if num >= -self.lower[0]:
                heapq.heappush(self.upper, num)
                return

            else:
                exchange = -heapq.heappop(self.lower)
                heapq.heappush(self.upper, exchange)
                heapq.heappush(self.lower, -num)
                return

        if len(self.upper) > len(self.lower):
            if num <= self.upper[0]:
                heapq.heappush(self.lower, -num)
                return
            else:
                exchange = heapq.heappop(self.upper)
                heapq.heappush(self.lower, -exchange)
                heapq.heappush(self.upper, num)
                return

    def findMedian(self) -> float:
        if len(self.upper) > len(self.lower):
            return self.upper[0]
        else:
            return (self.upper[0] + -self.lower[0]) / 2


# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()

if __name__ == "__main__":
    test = MedianFinder()
    test.addNum(40)
    print(test.findMedian())
    test.addNum(12)
    print(test.findMedian())
    test.addNum(16)
    print(test.findMedian())
    test.addNum(14)
    print(test.findMedian())
    test.addNum(35)
    print(test.findMedian())
    test.addNum(19)
    print(test.findMedian())
    test.addNum(34)
    print(test.findMedian())
    test.addNum(35)
    print(test.findMedian())
